### PR TITLE
Only emit metadata changed signals if we have to.

### PR DESCRIPTION
**Pull request by JasonLG1979**
Metadata is the most "expensive" property to deal with for clients. This cuts our Metadata updates to the bare minimum.

### CODE REVIEW

File: pithos/pithos.py [hunk 1]
```diff
@@ -834,17 +842,30 @@ def get_album_art(url, tmpdir, *extra):
...
+                self.update_song_row(song)
                 if file_url:
                     song.artUrl = file_url
-                    self.emit('metadata-changed', song)
-                self.update_song_row(song)
+                    if song is self.current_song:
+                        self.emit('metadata-changed', song)
+                # We tried to get covers for all the songs in the playlist,
+                # and the timeout is still live. Cancel it and emit
+                # a 'songs-added' signal.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/plugins/mpris.py [hunk 2]
```diff
@@ -383,13 +378,14 @@ def _metadatachange_handler(self, window, song):
...
-                if track_id == trackId:
+                if track_id == trackId and not self._metadata_equal(self._metadata_list[index], metadata):
                     self._metadata_list[index] = metadata
                     self.TrackMetadataChanged(trackId, metadata)
                     break
         # No need to update the current metadata if the current song has been banned
         # or set tired as it will be skipped anyway very shortly.
-        if song is window.current_song and not (song.tired or song.rating == 'ban'):
+        if (song is window.current_song and not (song.tired or song.rating == 'ban')
+            and not self._metadata_equal(self._metadata, metadata)):
```
Comment (TingPing) [inline pithos/plugins/mpris.py]: __[flake8]__
* [E129] visually indented line with same indent as next logical line
* [W503] line break before binary operator

File: pithos/plugins/mpris.py [hunk 3]
```diff
@@ -424,6 +420,20 @@ def _get_metadata(self, window, song):
 
         return metadata
 
+    def _metadata_equal(self, m1, m2):
+        # Test to see if 2 sets of metadata are the same
+        # to avoid unneeded updates.
+        if len(m1) != len(m2):
+            return False
```
Comment (TingPing) [inline pithos/plugins/mpris.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/plugins/mpris.py [hunk 4]
```diff
@@ -424,6 +420,20 @@ def _get_metadata(self, window, song):
 
         return metadata
 
+    def _metadata_equal(self, m1, m2):
```
Comment (TingPing) [inline pithos/plugins/mpris.py]: Shouldn't direct comparisons of dicts work just fine?, if GVariant doesn't implement comparisons that should probably be fixed upstream too.

Comment (JasonLG1979) [inline pithos/plugins/mpris.py]: No. They're GLib.Variants you have to unpack the values.

Comment (TingPing) [inline pithos/plugins/mpris.py]: PyGObject should be internally using [g_variant_compare()](https://developer.gnome.org/glib/stable/glib-GVariant.html#g-variant-compare)

Comment (JasonLG1979) [inline pithos/plugins/mpris.py]: I guess there's GLib.Variant.comapre. I'll test it out.

Comment (JasonLG1979) [inline pithos/plugins/mpris.py]: It says it only works on basic types. I may still need to unpack the artist.

Comment (TingPing) [inline pithos/plugins/mpris.py]: Sorry I meant `.equal()` rather than `.compare()` I think it can work on more types.

**Comment by JasonLG1979** [comment]
Something like that maybe?

**Comment by JasonLG1979** [comment]
@TingPing I'd really like to see this merged if you don't mind :)

**Pull request merged**